libraries/Networking/src/ESP_Network_Interface.cpp

```diff
@@ -0,0 +1,526 @@
+        _has_ip = false;
+#if ARDUHAL_LOG_LEVEL >= ARDUHAL_LOG_LEVEL_VERBOSE
+        log_v("%s Lost IP", desc());
+#endif
+#if SOC_WIFI_SUPPORTED
+        if(_interface_id == ESP_NETIF_ID_STA){
+            arduino_event.event_id = ARDUINO_EVENT_WIFI_STA_LOST_IP;
+            Network.clearStatusBits(STA_HAS_IP_BIT);
+        } else 
+#endif
+        // if(_interface_id == ESP_NETIF_ID_PPP){
+        //     arduino_event.event_id = ARDUINO_EVENT_PPP_LOST_IP;
+        //     Network.clearStatusBits(PPP_HAS_IP_BIT);
+        // } else 
+        if(_interface_id >= ESP_NETIF_ID_ETH && _interface_id < ESP_NETIF_ID_MAX){
+            arduino_event.event_id = ARDUINO_EVENT_ETH_LOST_IP;
+            Network.clearStatusBits(ETH_HAS_IP_BIT(_interface_id - ESP_NETIF_ID_ETH));
+        }
+    } else if(event_id == IP_EVENT_GOT_IP6){
+        ip_event_got_ip6_t* event = (ip_event_got_ip6_t*) event_data;
+        esp_ip6_addr_type_t addr_type = esp_netif_ip6_get_addr_type(&event->ip6_info.ip);
+#if ARDUHAL_LOG_LEVEL >= ARDUHAL_LOG_LEVEL_VERBOSE        
+        char if_name[NETIF_NAMESIZE] = {0,};
+        netif_index_to_name(event->ip6_info.ip.zone, if_name);
+        static const char * addr_types[] = { "UNKNOWN", "GLOBAL", "LINK_LOCAL", "SITE_LOCAL", "UNIQUE_LOCAL", "IPV4_MAPPED_IPV6" };
+        log_v("IF %s Got IPv6: Interface: %d, IP Index: %d, Type: %s, Zone: %d (%s), Address: " IPV6STR, desc(), _interface_id, event->ip_index, addr_types[addr_type], event->ip6_info.ip.zone, if_name, IPV62STR(event->ip6_info.ip));
+#endif
+        memcpy(&arduino_event.event_info.got_ip6, event_data, sizeof(ip_event_got_ip6_t));
+#if SOC_WIFI_SUPPORTED
+        if(_interface_id == ESP_NETIF_ID_STA){
+            arduino_event.event_id = ARDUINO_EVENT_WIFI_STA_GOT_IP6;
+            if(addr_type == ESP_IP6_ADDR_IS_GLOBAL){
+                Network.setStatusBits(STA_HAS_IP6_GLOBAL_BIT);
+            } else if(addr_type == ESP_IP6_ADDR_IS_LINK_LOCAL){
+                Network.setStatusBits(STA_HAS_IP6_BIT);
+            }
+        } else if(_interface_id == ESP_NETIF_ID_AP){
+            arduino_event.event_id = ARDUINO_EVENT_WIFI_AP_GOT_IP6;
+            if(addr_type == ESP_IP6_ADDR_IS_LINK_LOCAL){
+                Network.setStatusBits(AP_HAS_IP6_BIT);
+            }
+        } else 
+#endif
+        // if(_interface_id == ESP_NETIF_ID_PPP){
+        //     arduino_event.event_id = ARDUINO_EVENT_PPP_GOT_IP6;
+        //     Network.setStatusBits(PPP_HAS_IP6_BIT);
+        // } else 
+        if(_interface_id >= ESP_NETIF_ID_ETH && _interface_id < ESP_NETIF_ID_MAX){
+            arduino_event.event_id = ARDUINO_EVENT_ETH_GOT_IP6;
+            if(addr_type == ESP_IP6_ADDR_IS_GLOBAL){
+                Network.setStatusBits(ETH_HAS_IP6_GLOBAL_BIT(_interface_id - ESP_NETIF_ID_ETH));
+            } else if(addr_type == ESP_IP6_ADDR_IS_LINK_LOCAL){
+                Network.setStatusBits(ETH_HAS_IP6_BIT(_interface_id - ESP_NETIF_ID_ETH));
+            }
```

Comment:
This list is incomplete; addr_type has some other values, including ULA (unique local address), which are also global scope. The ESP definition of each category is here: https://github.com/espressif/esp-lwip/blob/2.1.2-esp/src/include/lwip/ip6_addr.h#L203

Note that their definition of GLOBAL is not exactly the same as the current definition global scope in RFC 4291 (ULAs have global scope), but is probably the best name for the range which is the only current range used for global assignment. See: https://www.iana.org/assignments/ipv6-address-space/ipv6-address-space.xhtml

While other types (e.g. site local) are unlikely, ULA addresses could be found in private networks (they are similar to IPv4 private address ranges), including as DNS destinations. They could also be used for NAT66 (like private address range NAT44), so when in use may also expect to reach public DNS locations (via NAT).